Python/pathconfig.c

```diff
@@ -18,6 +18,154 @@
 extern "C" {
 #endif
 
+#ifdef __MINGW32__
+#define wcstok  wcstok_s
+#include <windows.h>
+#endif
+
+static int
+Py_StartsWithA(const char * str, const char * prefix)
+{
+    while(*prefix)
+    {
+        if(*prefix++ != *str++)
+            return 0;
+    }
+
+    return 1;
+}
+
+static int
+Py_StartsWithW(const wchar_t * str, const wchar_t * prefix)
+{
+    while(*prefix)
+    {
+        if(*prefix++ != *str++)
+            return 0;
+    }
+
+    return 1;
+}
+
+char
+Py_GetSepA(const char *name)
+{
+    static char sep = '\0';
+#ifdef _WIN32
+    /* https://msdn.microsoft.com/en-gb/library/windows/desktop/aa365247%28v=vs.85%29.aspx
+     * The "\\?\" prefix .. indicate that the path should be passed to the system with minimal
+     * modification, which means that you cannot use forward slashes to represent path separators
+     */
+    if (name != NULL && Py_StartsWithA(name, "\\\\?\\") != 0)
+    {
+        return '\\';
+    }
+#endif
+    if (sep != '\0')
+        return sep;
+#if defined(__MINGW32__)
+    char* msystem = getenv("MSYSTEM");
+    if (msystem != NULL && strcmp(msystem, "") != 0)
+        sep = '/';
+    else
+        sep = '\\';
+#endif
+    return SEP;
```

Comment:
I'm having trouble wrapping my head around this, but I think that this could return the "wrong" slash on the first call in `__MINGW32__` case.  Maybe something like
```suggestion
#else
    sep = SEP;
#endif
    return sep;
```

Also, there used to be a rule some C compilers enforced about variable definitions coming before other statements, I don't know if the Python .c files still follow this rule, but that `char* msystem` would now violate it where it didn't before.  The only compiler I ever dealt with that normally rejected such code was an ancient version of MSVC though.

Reply:
> I'm having trouble wrapping my head around this, but I think that this could return the "wrong" slash on the first call in __MINGW32__ case. Maybe something like

I'll update it, thanks!

> Also, there used to be a rule some C compilers enforced about variable definitions coming before other statements, I don't know if the Python .c files still follow this rule, but that char* msystem would now violate it where it didn't before. The only compiler I ever dealt with that normally rejected such code was an ancient version of MSVC though.

I don't think we need to care about that, since this is specific to msys2.